swap/protocol_test.go

```diff
@@ -0,0 +1,317 @@
+
+	log.Debug("create peer instances")
+	// create Peer instances
+	// NOTE: remember that these are peer instances representing each **a model of the remote peer** for every local node
+	// so creditor is the model of the remote mode for the debitor! (and vice versa)
+
+	// in order to be able to model as realistically as possible sending and receiving, let's use a MsgPipe
+	// a MsgPipe is a duplex read-write object, write to one end and read from the other
+
+	// create the message pipe
+	crw, drw := p2p.MsgPipe()
+	// create the creditor peer
+	cPtpPeer := p2p.NewPeer(enode.ID{}, "creditor", []p2p.Cap{})
+	cProtoPeer := protocols.NewPeer(cPtpPeer, crw, Spec)
+	// create the debitor peer
+	dPtpPeer := p2p.NewPeer(enode.ID{}, "dreditor", []p2p.Cap{})
+	dProtoPeer := protocols.NewPeer(dPtpPeer, drw, Spec)
+	// create the Swap protocol peers
+	creditor := NewPeer(cProtoPeer, debitorSwap, debitorSwap.backend, creditorSwap.owner.address, debitorSwap.owner.Contract)
+	debitor := NewPeer(dProtoPeer, creditorSwap, creditorSwap.backend, debitorSwap.owner.address, debitorSwap.owner.Contract)
+
+	// set balance artificially
+	creditorSwap.balances[debitor.ID()] = 42
+	log.Debug("balance", "balance", creditorSwap.balances[debitor.ID()])
+	// a safe check: at this point no cheques should be in the swap
+	if len(creditorSwap.cheques) != 0 {
+		t.Fatalf("Expected no cheques at creditor, but there are %d:", len(creditorSwap.cheques))
+	}
+
+	log.Debug("create a cheque")
+	var err error
+	cheque := &Cheque{
+		ChequeParams: ChequeParams{
+			Contract:    debitorSwap.owner.Contract,
+			Beneficiary: creditorSwap.owner.address,
+			Amount:      42,
+			Honey:       42,
+			Timeout:     0,
+		},
+	}
+	cheque.Signature, err = debitorSwap.signContent(cheque)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	emitMsg := &EmitChequeMsg{
+		Cheque: cheque,
+	}
+	log.Debug("send the message with the cheque to the beneficiary")
+	go creditor.Send(ctx, emitMsg)
+
+	log.Debug("read the message on the beneficiary through the pipe")
+	msg, err := drw.ReadMsg()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	log.Debug("convert the message to our message type (simulated p2p.protocols)")
+	var wmsg protocols.WrappedMsg
+	err = msg.Decode(&wmsg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	msg.Discard()
+
+	val, ok := Spec.NewMsg(msg.Code)
+	if !ok {
+		t.Fatalf("invalid message code: %v", msg.Code)
+	}
+	if err := rlp.DecodeBytes(wmsg.Payload, val); err != nil {
+		t.Fatalf("decode error <= %v: %v", msg, err)
+	}
+
+	log.Debug("trigger reading the message on the beneficiary")
+
+	err = debitor.handleEmitChequeMsg(ctx, val.(*EmitChequeMsg))
+	if err != nil {
+		t.Fatal(err)
+	}
+	log.Debug("balance", "balance", creditorSwap.balances[debitor.ID()])
+	// check that the balance has been reset
+	if creditorSwap.balances[debitor.ID()] != 0 {
+		t.Fatalf("Expected debitor balance to have been reset to %d, but it is %d", 0, creditorSwap.balances[debitor.ID()])
+	}
+	/*
```

Comment:
remove this or deal with it :)

Reply:
We can't deal with it now, because it would make the PR even more bigger. There is development work to be done for this to work. It is one of the highers priority tasks for me as soon as this PR is merged. And it will be done right after it. And as soon as it is done, I will want to activate this code, because it is the ultimate test condition for this test. 

Thus I would really think it should stay. It will go away very soon. It is a reminder that the test is not complete as-is.

Reply:
ok